cmd/dumpstate/dumpstate.go

```diff
@@ -0,0 +1,238 @@
+	Source      weave.Address  `json:"source"`
+	Arbiter     weave.Address  `json:"arbiter"`
+	Destination weave.Address  `json:"destination"`
+	Timeout     weave.UnixTime `json:"timeout"`
+	Amount      []*coin.Coin   `json:"amount"`
+	Address     weave.Address  `json:"address"`
+}
+
+type contractFormat struct {
+	Participants        []*multisig.Participant `json:"participants"`
+	ActivationThreshold multisig.Weight         `json:"activation_threshold"`
+	AdminThreshold      multisig.Weight         `json:"admin_threshold"`
+	Address             weave.Address           `json:"address"`
+}
+
+func main() {
+	flag.Usage = func() {
+		fmt.Fprintln(flag.CommandLine.Output(), `
+Export state data. Pipe-in app version as input.`)
+		flag.PrintDefaults()
+	}
+	var (
+		dbFl = flag.String("db", env("BNSD_DB_DIR", os.ExpandEnv("$HOME")+"/.bns"),
+			"bnsd database directory")
+		heightFl = flag.Uint("height", 0,
+			"commit height")
+		outFl = flag.String("out", "./dump.json",
+			"dump output directory")
+	)
+
+	flag.Parse()
+
+	if *heightFl == 0 {
+		flag.PrintDefaults()
+		os.Exit(1)
+	}
+
+	// validate db
+	dbPath := filepath.Join(*dbFl, "bns.db")
+	_, err := os.Stat(dbPath)
+	if os.IsNotExist(err) {
+		fmt.Printf("db file does not exists: %s\n", err)
+		os.Exit(1)
+	}
+
+	// create db store
+	kv, err := bnsd.CommitKVStore(dbPath)
+	if err != nil {
+		fmt.Printf("cannot initialize bnsd commit store: %s\n", err)
+		os.Exit(1)
+	}
+
+	store := app.NewCommitStore(kv)
+	// set db version/height
+	err = kv.LoadVersion(int64(*heightFl))
+	if err != nil {
+		fmt.Printf("cannot load db version: %s\n", err)
+		os.Exit(1)
+	}
+
+	// create output file
+	outFile, err := os.Create(*outFl)
+	if err != nil {
+		fmt.Printf("cannot create output file: %s\n", err)
+		os.Exit(1)
+	}
+
+	usernames, err := extractUsername(store)
+	if err != nil {
+		fmt.Printf("cannot extract usernames: %s\n", err)
+		os.Exit(1)
+	}
+	escrows, err := extractEscrow(store)
+	if err != nil {
+		fmt.Printf("cannot extract escrows: %s\n", err)
+		os.Exit(1)
+	}
+	contracts, err := extractContracts(store)
+	if err != nil {
+		fmt.Printf("cannot extract contracts: %s\n", err)
+		os.Exit(1)
+	}
+
+	wallets, err := extractWallets(store)
+	if err != nil {
+		fmt.Printf("cannot extract wallets: %s\n", err)
+		os.Exit(1)
+	}
+
+	outJson := Out{
+		Username: usernames,
+		Escrow:   escrows,
+		Contract: contracts,
+		Wallets:  wallets,
+	}
+	err = json.NewEncoder(outFile).Encode(outJson)
+	if err != nil {
+		fmt.Printf("cannot write to file: %s\n", err)
+		os.Exit(1)
+	}
+}
+
+func extractUsername(store *app.CommitStore) ([]tokenFormat, error) {
+	it := orm.IterAll("tokens")
+	var out []tokenFormat
+	for {
+		var token username.Token
+		switch key, err := it.Next(store.CheckStore(), &token); {
+		case err == nil:
+			out = append(out, tokenFormat{
+				Username: string(key),
+				Targets:  token.Targets,
+				Owner:    token.Owner,
+			})
+		case errors.ErrIteratorDone.Is(err):
+			goto success
```

Comment:
The use of `goto` is very uncommon in Go projects. Why not `return out, nil` here to exit early? You could also break with a label if you have to. 
Disclaimer: I did not review the PR. Just stumbled upon the `goto`.